### ISSUE TITLE
TupleVariation.hasImpact()

### ISSUE THREAD

**Issue by behdad** [fonttools/fonttools#2263]
That function is:
```py
        def hasImpact(self):
                """Returns True if this TupleVariation has any visible impact.

                If the result is False, the TupleVariation can be omitted from the font
                without making any visible difference.
                """
                return any(c is not None for c in self.coordinates)
```

However, a TV that has only zeros (zero coord (0,0) in the case of gvar; 0 in case of cvar) also has no visible impact. So I'm confused why this function was written this way.

Moreover, currently `compileTupleVariationStore` starts by filtering out tuples that have no impact:
```
def compileTupleVariationStore(variations, pointCount,
                               axisTags, sharedTupleIndices,
                               useSharedPoints=True):
        variations = [v for v in variations if v.hasImpact()]
```

I like to remove that check. The creator of TV should do the check, and indeed varLib does.

One of the tests fails. But then again, that's exactly testing this behavior, not something that a reasonable client would do IMO.  I don't even know how you'd get all None... The IUP optimize generates that. But as I said, if we are filtering these out, then:
1. we should also filter out all-zero deltas,
2. ideally this should be done earlier, such that the sharedCoordinates calculations are not affected.

In summary, not a big deal the way it is. Is just a weird half-check.

WDYT?

**Comment by anthrotype** [comment]
> The IUP optimize generates that.

yes, and `iup_delta_optimize` is used not only in `varLib._add_gvar` but also in `TupleVariation.optimize` method (the code is also unnecessarily duplicated). True, the `_add_gvar` skips adding tuples for which `all(v == 0 for v in delta.array)`.
The `TupleVariation.optimize` method is used by the `varLib.instancer` in `instantiateGvarGlyph` function, to re-optimize the tuples that are left in the partial instance (these need to be inferred in order merge tuples with overlapping tents):

https://github.com/fonttools/fonttools/blob/cbb33d8aa9e21f01ba72700d5748d16c09e0342b/Lib/fontTools/varLib/instancer/__init__.py#L360-L363

I guess it is possible that the iup-optimize leaves behind a noop tuple with all Nones. In which case it would be dropped upon compile thanks to `hasImpact`.

So if you decide to remove that check, and leave to the caller to decide what to do with all-zeros deltas, make sure the instancer drops the no-op tuples.
